src/garage/sampler/rl2_sampler.py

```diff
@@ -0,0 +1,195 @@
+"""BatchSampler which uses VecEnvExecutor to run multiple environments."""
+from collections import OrderedDict
+import itertools
+import pickle
+import time
+
+from dowel import logger, tabular
+import numpy as np
+
+from garage.experiment import deterministic
+from garage.misc import tensor_utils
+from garage.misc.prog_bar_counter import ProgBarCounter
+from garage.sampler.batch_sampler import BatchSampler
+from garage.sampler.stateful_pool import singleton_pool
+from garage.sampler.utils import truncate_paths
+from garage.sampler.vec_env_executor import VecEnvExecutor
+
+
+class RL2Sampler(BatchSampler):
+    """BatchSampler which uses VecEnvExecutor to run multiple environments.
+
+    This sampler is for RL^2. See https://arxiv.org/pdf/1611.02779.pdf.
+
+    Args:
+        algo (garage.np.algos.RLAlgorithm): An algorithm instance.
+        env (garage.envs.GarageEnv): An environement instance.
+        meta_batch_size (int): Number of meta batch for sampling.
+        episode_per_task (int): Number of episode for each task.
+        n_envs (int): Number of environment instances to setup.
+            This parameter has effect on sampling performance.
+
+    """
+
+    def __init__(self,
+                 algo,
+                 env,
+                 meta_batch_size,
+                 episode_per_task,
+                 n_envs=None):
+        if n_envs is None:
+            n_envs = singleton_pool.n_parallel * 4
+        super().__init__(algo, env)
+        self._n_envs = n_envs
+
+        self._meta_batch_size = meta_batch_size
+        self._episode_per_task = episode_per_task
+        self._vec_env = None
+        self._env_spec = self.env.spec
+
+    def start_worker(self):
+        """Start workers."""
+        n_envs = self._n_envs
+        assert n_envs >= self._meta_batch_size, (
+            'Number of vectorized environments '
+            'should be at least meta_batch_size.')
```

Comment:
why? if it's less than meta-batch size can't i just run them in series?

Reply:
How should we handle the allocation of environments over tasks If we don't enforce it with some constraints? e.g. meta_batch_size = 5 but there are 3 parallel envs.

Reply:
i agree this is challenging within this implementation, but does it have to be this limiting?

what would you do if meta_batch_size was 100 but you can only have 10 CPUs for workers?

Reply:
10 envs at a time, for 10 times. How about limiting `n_workers % meta_batch_size == 0` if `n_workers < meta_batch_size`, and `meta_batch_size % n_workers == 0`  if `meta_batch_size > n_workers`?